### ISSUE TITLE
Case sensitivity

### ISSUE THREAD

**Issue by kmcphillips** [Shopify/measured#3]
For example:

```ruby
unit?("KG")
Measured::UnitError: Cannot find unit for KG.
```

At the very least this should return false and not raise.

**Comment by AnthonyBobsin** [comment]
@kmcphillips I am thinking up an implementation for this issue. Originally, I thought it would be nice to convert all units to lowercase, and do the same for the search string. Although that would lead to unpredictability, and some limitations in terms of unit naming. I think it would be best if, like you said, it just returns false. Let me know if you have any other ideas.

**Comment by kmcphillips** [comment]
Well. Downcasing would be the obvious solution. 

I mean, each implementation of a system of measure defines its list of aliases.
https://github.com/Shopify/measured/blob/master/lib/measured/units/length.rb

If `KG` is common it can just be added there. But then what about `Kilogram` then `KiloGram` `KILOGRAM` etc.. Seems unsustainable.

A case sensitivity option could be used:

```ruby
  conversion.add :cm,
    aliases: [:centimeter, :centimetre, :centimeters, :centimetres],
    case_sensitive: false,
    value: "0.01 m"
```

Have an opinion?

**Comment by kmcphillips** [comment]
As for the `false` vs raise we have been talking about that quite a bit. As we are using the rails adapter it is undesirable to have a situation where initializing a measurement from user input raises. That usually does not happen. But returning `nil` or `false` or something is not clear either.

**Comment by AnthonyBobsin** [comment]
I really like that case_sensitive option. The only reason I was against downcasing all checks, was if a user was to add their own units, they might be confused as to why they cannot create two different units with the same name but different cases. It is a very small edge case, but just in case. 

I will work on implementing the case_sensitive option, then anything past that, it should be sufficient for the user to create their own conversion with custom aliases. Additionally, I can work on something that adds the option for the user to add aliases to existing units.

**Comment by kmcphillips** [comment]
Right. Because right now it's a single definition. Adding the ability to extend it would be great.

If you want to jump on some of this, power to you! The DSL is mostly defined here:
https://github.com/Shopify/measured/blob/96ae34672d44ba21e73e218ce9fbd1770bebeb7b/lib/measured/conversion.rb#L10-16

But each unit is an instance of `Measured::Unit` defined here:
https://github.com/Shopify/measured/blob/96ae34672d44ba21e73e218ce9fbd1770bebeb7b/lib/measured/unit.rb

So it should be a matter of adding methods to add aliases to an instance of unit, and maybe a convenience method in the DSL to do so.

But I'm totally fine with adding the `case_sensitive` option to the DSL and setting it false for weights and lengths.

**Comment by kmcphillips** [comment]
The change we want is in here somewhere:

```ruby
[5] pry(length):1> Measured::Length.conversion.units
=> [#<Measured::Unit: m (m, meter, meters, metre, metres) >,
 #<Measured::Unit: cm (centimeter, centimeters, centimetre, centimetres, cm) 0.01 m>,
 #<Measured::Unit: mm (millimeter, millimeters, millimetre, millimetres, mm) 0.001 m>,
 #<Measured::Unit: in (in, inch, inches) 0.0254 m>,
 #<Measured::Unit: ft (feet, foot, ft) 0.3048 m>,
 #<Measured::Unit: yd (yard, yards, yd) 0.9144 m>]
[6] pry(length):1> Measured::Length.conversion.unit_for(:cm)
NoMethodError: private method `unit_for' called for #<Measured::Conversion:0x007f02f90ce4e8>
from (pry):4:in `__binding__'
```

**Comment by AnthonyBobsin** [comment]
I'll gladly hop on this issue. Perfect, ya I believe I already have a general idea of how to go about this. Thanks for all your help with pointing me in the right direction! I'll work on this on some free time, and submit a merge request when complete.

**Comment by kmcphillips** [comment]
Amazing!

Don't hesitate to ask for help. We are actively building and implementing this tool as we are using it more, so there are still changes coming. Thanks for helping us improve it.

**Comment by AnthonyBobsin** [comment]
Sounds good, will do! Anytime, love the tool, just happy I'm able to help out

**Comment by AnthonyBobsin** [comment]
I'm just about wrapped up with the implementation, just have a couple more tests to write. Before I submit a merge request I just wanted to clarify some things about case_sensitive flag defaults.

Currently I have it so that any Measured::Unit gets initialized with case_sensitive set to true by default. All the existing units that we supply through the gem all have case_sensitive set to false.

Is this how you want it? The major part I'm questioning is what case_sensitive should be set to by default in initialization.

**Comment by kmcphillips** [comment]
I'd say case sensitivity would be off by default.

**Comment by kmcphillips** [comment]
closed by #20
